google/cluster/configuration.tf

```diff
@@ -56,4 +56,9 @@ locals {
   manifest_path         = var.manifest_path != null ? var.manifest_path : local.manifest_path_default
 
   disable_default_ingress = lookup(local.cfg, "disable_default_ingress", false)
+
+  enable_private_nodes    = lookup(local.cfg, "enable_private_nodes", true)
+  enable_private_endpoint = lookup(local.cfg, "enable_private_endpoint", false)
+  master_cidr_block       = lookup(local.cfg, "master_cidr_block", "10.254.0.0/28")
```

Comment:
Why this default? Is there a default on Google's side or in the Terraform provider? If so, I'd prefer to set this to `null` here and use an upstream default.

I'd also prefer if we could keep the name equally to the upstream name here, so `master_ipv4_cidr_block` including the ipv4 part. I can only assume Google did this to be prepared for a potential ipv6 version of this attribute.

Reply:
This is required and there is no default, unfortunately, hence why I tried to use a range that hopefully nobody would use. also when it comes to terraform variables you cant use numbers in the name (hence why I removed the `ipv4` part)

Reply:
We should split these two topics. Mixing in this one PR adds quite a bit of complexity. Especially given the upstream bug.

I suggest we leave this PR for private nodes only, and hardcode enable_private_endpoint to false. If required we can then add support for private endpoints in another PR where we can figure out what the right approach for the CIDR is. For what it's worth the [GKE docs use a 172 type address](https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters). I have not looked into this enough to understand the implications of any default value.

Reply:
The problem is that the `master_ipv4_cidr_block ` is required when you set `master_ipv4_cidr_block ` to `true` not only `enable_private_endpoint`. I can look at updating it to the `172` range

Reply:
Just a note on this, I removed the `enable_private_endpoint ` as to my understanding this will put the API on a private network which will mean Terraform can not manage the resources within (which will mean Kubestacks manifests will not work without something akin to a bastion host)

Reply:
No what I mean is, private nodes and private API are two separate features. Yes you may need to set both variables. But you don't have to expose it as variables outside the module. Simply make this PR only about private nodes so we can sort the bug out with upstream and find a way how to provide a migration path to users.

We leave private API for another PR later.

I hadn't read up on this before. But I did now and it seems to me, the way this works is it requires the new VPC-native cluster mode to not have public IPs for nodes. That's why `ip_allocation_policy` is required. Previously created clusters are routes-based. From what I found so far, there is no way to migrate a cluster from routes to VPC-nate without downtime.

Reply:
It seems the default was changed for UI and CLI created clusters, but not for the Terraform module. I suppose this is due to no other backwards compatible option?

| In both gcloud and the Cloud Console, VPC-native is the default for new clusters and increasingly, GKE features such as Standalone Network Endpoint Groups (NEGs) have relied on clusters being VPC-native. In Terraform however, the default behaviour is to create a routes-based cluster for backwards compatibility.

Source: https://www.terraform.io/docs/providers/google/guides/using_gke_with_terraform.html#vpc-native-clusters

Reply:
To move this forward I suggest you set default values that match what's in the Google docs but make the CIDRs configurable. Similar to https://github.com/kbst/terraform-kubestack/pull/135/commits/f6efd01f81fbc0a2c814e1a1f9055f6ae420ba78, so it's similar between the GKE and AKs modules. What do you think?

Reply:
What do you think about the above @Spazzy757?

Reply:
Completely missed this, I can look at doing this tomorrow! I like it